### PR TITLE
add script for lowres according to fullres format

### PR DESCRIPTION
**Pull request by deep-diver**
uploaded the dataset to `gs://sidewalks-tfx-lowres/sidewalks-tfrecords`

maybe, we could merge this script with `create_tfrecords_fullres.py` by giving a option `--resize`.

### CODE REVIEW

File: tfrecords/create_tfrecords_lowres.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    ds = ds["train"].train_test_split(test_size=args.split, seed=args.seed)
+    train_ds = ds["train"]
+    val_ds = ds["test"]
+
+    return train_ds, val_ds
+
+
+def normalize_img(image: np.ndarray, label: np.ndarray) -> Tuple[tf.Tensor, tf.Tensor]:
+    image = tf.cast(image, tf.float32) / 255.0
+    label -= 1
```
Comment (sayakpaul) [inline tfrecords/create_tfrecords_lowres.py:45]: Is this needed in our case? I know it's there in https://github.com/deep-diver/semantic-segmentation-ml-pipeline/blob/main/tfrecords/create_tfrecords_fullres.py too. But I guess it's not needed.

**Review by sayakpaul — APPROVED** [review]
I agree with your suggestion of merging the two scripts. Let's add that in this PR. 

After generating the TFRecords, let's also verify them with the TFRecord parsing notebook.

File: tfrecords/create_tfrecords_lowres.py (hunk 1 above)
Comment (sayakpaul) [reply tfrecords/create_tfrecords_lowres.py:45]: Yeah I mean we cal eliminate that step 'cause it seems like it's redundant. I remember it actually led to a training fault?

**Comment by deep-diver** [comment]
@sayakpaul 

addressed your comment! 
also, modified TFRecords for the low-resolution dataset are fully uploaded to the GCS

**Comment by deep-diver** [comment]
@sayakpaul 

also modify the existing notebook to show two different versions of TFRecords.

File: tfrecords/create_tfrecords.py [hunk 2]
```diff
@@ -41,51 +40,56 @@ def load_sidewalks_dataset(args):
     return train_ds, val_ds
 
 
-def normalize_img(image: np.ndarray, label: np.ndarray) -> Tuple[tf.Tensor, tf.Tensor]:
-    image = tf.cast(image, tf.float32) / 255.0
-    return image, label
```
Comment (sayakpaul) [inline tfrecords/create_tfrecords.py:46]: Why is this deleted?

File: tfrecords/create_tfrecords.py [hunk 3]
```diff
@@ -11,7 +11,7 @@
 
 References:
 
-    * https://github.com/sayakpaul/TF-2.0-Hacks/blob/master/Cats_vs_Dogs_TFRecords.ipynb
+    * https://github.com/GoogleCloudPlatform/practical-ml-vision-book/blob/master/05_create_dataset/05_split_tfrecord.ipynb
     * https://www.tensorflow.org/tutorials/images/segmentation
 """
```
Comment (sayakpaul) [inline tfrecords/create_tfrecords.py:16]: Since this script can handle two different things (full-res and resized res), let's add a usage.

File: tfrecords/create_tfrecords_str.py [hunk 4]
```diff
@@ -11,7 +11,7 @@
 
 References:
```
Comment (sayakpaul) [inline tfrecords/create_tfrecords_str.py:12]: What's the purpose of this script?

**Review by sayakpaul — CHANGES_REQUESTED** [review]

File: notebooks/parse_tfrecords.ipynb [hunk 5]
```diff
@@ -1,168 +1,250 @@
 {
```
Comment (sayakpaul) [inline notebooks/parse_tfrecords.ipynb:1]: <p>Didn't we have to apply transposition trick to make it work? I remember we ran into errors without it. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/deep-diver/semantic-segmentation-ml-pipeline/pull/28/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

File: tfrecords/create_tfrecords_str.py (hunk 4 above)
Comment (deep-diver) [reply tfrecords/create_tfrecords_str.py:12]: this is a reference script to create string-base encoded TFRecord

File: tfrecords/create_tfrecords.py (hunk 2 above)
Comment (deep-diver) [reply tfrecords/create_tfrecords.py:46]: since I am going to apply `mobilenet_v2.preprocess_input` to the input dataset, I guess it does perform normalization

File: tfrecords/create_tfrecords.py (hunk 3 above)
Comment (deep-diver) [reply tfrecords/create_tfrecords.py:16]: sure thing!

File: notebooks/parse_tfrecords.ipynb (hunk 5 above)
Comment (deep-diver) [reply notebooks/parse_tfrecords.ipynb:1]: I think we needed it to let TF know the shape of the Tensors, so I will apply `reshape` in the preprocessing step

Comment (sayakpaul) [reply notebooks/parse_tfrecords.ipynb:1]: Okay. Cool.

File: tfrecords/create_tfrecords_str.py (hunk 4 above)
Comment (sayakpaul) [reply tfrecords/create_tfrecords_str.py:12]: I see `tf.io.serialize_tensor()` being used. Why was it needed?

File: tfrecords/create_tfrecords.py (hunk 2 above)
Comment (sayakpaul) [reply tfrecords/create_tfrecords.py:46]: Then let's apply that here during the creation of TFRecords?

Comment (deep-diver) [reply tfrecords/create_tfrecords.py:46]: Possible.

But just wanted to let readers know it is possible to do so in Transform since this project focuses on ML pipeline not the dataset itself. WDYT?

File: tfrecords/create_tfrecords_str.py (hunk 4 above)
Comment (deep-diver) [reply tfrecords/create_tfrecords_str.py:12]: This is a script named as 'create_tfrecords.py' previously. I just renamed it as '_str'. You want me te delete this?

Comment (sayakpaul) [reply tfrecords/create_tfrecords_str.py:12]: Oh got it. Sorry, I got confused. Not let's keep it but also indicate the reason why the script exists.

File: tfrecords/create_tfrecords.py (hunk 2 above)
Comment (sayakpaul) [reply tfrecords/create_tfrecords.py:46]: Makes perfect sense to me.

**Comment by deep-diver** [comment]
@sayakpaul 

please check out the `tfrecords/README.md` for the updated description regarding TFRecords.

File: tfrecords/README.md [hunk 6]
```diff
@@ -1,2 +1,21 @@
-* `create_tfrecords.py` creates TFRecords with a reduced spatial size so that we don't need to use Dataflow within a Vertex AI pipeline. 
-* `create_tfrecords_fullres.py` creates TFRecords using the original spatial size. 
\ No newline at end of file
+There are two scripts to create TFRecords for [sidewalk dataset](https://huggingface.co/datasets/segments/sidewalk-semantic) hosted in 🤗 Dataset Hub. They are basically same except for one thing. `create_tfrecords.py` creates TFRecords with the features encoded in numeric data type. On the other hand, `create_tfrecords_str.py` creates TFRecords with the features encoded in string data type. 
+- This is due to demonstrate how to create such dataset because string encoding is better in terms of compression.
```
Comment (sayakpaul) [inline tfrecords/README.md:2]:
```suggestion
- This is to demonstrate how to create such a dataset because string encoding is better in terms of compression.
```

File: tfrecords/README.md [hunk 7]
```diff
@@ -1,2 +1,21 @@
...
+we created two different datasets with the following CLIs:
+```console
+# full resolution dataset
+$ python create_tfrecords.py
+
+# lowered resolution dataset
+$ python create_tfrecords.py --resize 256
+```
+
+In Google Cloud environment, the default machine couldn't not handle a large size of dataset, so we make `256` sized dataset. The original size dataset is there to experiment that how `ImportExampleGen` and `Transform` components could delegate their jobs to **Dataflow**.
```
Comment (sayakpaul) [inline tfrecords/README.md:21]: Even without `Transform` we can use `ImportExampleGen` with Dataflow to integrate with the full resolution TFRecords right?

File: tfrecords/create_tfrecords_str.py [hunk 8]
```diff
@@ -41,45 +43,43 @@ def load_sidewalks_dataset(args):
 
 def normalize_img(image: np.ndarray, label: np.ndarray) -> Tuple[tf.Tensor, tf.Tensor]:
     image = tf.cast(image, tf.float32) / 255.0
-    label -= 1
+    return image, label
```
Comment (sayakpaul) [inline tfrecords/create_tfrecords_str.py:46]: I suggest we remove it from here too so as to delegate the preprocessing logic to `Transform`.

File: tfrecords/README.md (hunk 6 above)
Comment (sayakpaul) [reply tfrecords/README.md:2]: Let's also add a note that if the images are serialized as strings, their schema for decoding / parsing needs to change accordingly. [This guide](https://www.tensorflow.org/tutorials/load_data/tfrecord) provides a good overview.

**Review by sayakpaul — CHANGES_REQUESTED** [review]

File: tfrecords/README.md (hunk 7 above)
Comment (deep-diver) [reply tfrecords/README.md:21]: that is correct. this sentence means `Transform` could be run on Dataflow as well. I think `Transform` won't be able to handle such big data if `ImportExampleGen` can't, so I guessed we need to enable Dataflow for not only `ImportExampleGen` but also `Transform`.

Comment (sayakpaul) [reply tfrecords/README.md:21]: Let's maybe clarify it fully then?

**Comment by deep-diver** [comment]
@sayakpaul 

addressed your comments :)

**Review by sayakpaul — APPROVED** [review]
Great work!
